File: ios/Libraries/DDPClient.swift
import Foundation

/// Minimal DDP WebSocket client for listening to Rocket.Chat media-signal events from native iOS.
/// Only implements the subset needed to detect call hangup: connect, login, subscribe, and ping/pong.
final class DDPClient {
    private struct QueuedMethodCall {
        let method: String
        let params: [Any]
        let completion: (Bool) -> Void
    }
    
    private static let TAG = "RocketChat.DDPClient"
    private let stateQueueKey = DispatchSpecificKey<Void>()
    private let stateQueue = DispatchQueue(label: "chat.rocket.reactnative.ddp-client")
    
    private var webSocketTask: URLSessionWebSocketTask?
    private var urlSession: URLSession?
    private var sendCounter = 0
    private var isConnected = false
    
    /// Called for every incoming DDP collection message (e.g. stream-notify-user).
    var onCollectionMessage: (([String: Any]) -> Void)?

    init() {
        stateQueue.setSpecific(key: stateQueueKey, value: ())
    }

    deinit {
        disconnectOnStateQueue(error: nil)
    }

    // MARK: - Connect
    
    func connect(host: String, completion: @escaping (Bool) -> Void) {
        stateQueue.async {
            let wsUrl = Self.buildWebSocketURL(host: host)
            
            guard let url = URL(string: wsUrl) else {
                #if DEBUG
                print("[\(Self.TAG)] Invalid WebSocket URL: \(wsUrl)")
                #endif
                completion(false)
                return
            }
            
            #if DEBUG
            print("[\(Self.TAG)] Connecting to \(wsUrl)")
            #endif

            // Cancel any existing connection before creating a new one.
            self.webSocketTask?.cancel(with: .normalClosure, reason: nil)
            self.urlSession?.invalidateAndCancel()
            self.webSocketTask = nil
            self.urlSession = nil

            let session = URLSession(configuration: .default, delegate: DDPClientURLSessionChallengeDelegate(), delegateQueue: nil)
            let task = session.webSocketTask(with: url)

            self.urlSession = session
            self.webSocketTask = task
            self.isConnected = false
            task.resume()
            
            self.listenForMessages(task: task)
            
            let connectMsg: [String: Any] = [
                "msg": "connect",
                "version": "1",
                "support": ["1", "pre2", "pre1"]
            ]
            
            self.send(connectMsg) { [weak self] success in
                guard let self else { return }
                if success {
                    self.waitForConnected(timeout: 10.0, completion: completion)
                } else {
                    completion(false)
                }
            }
        }
    }
    
    // MARK: - Login
    
    func login(token: String, completion: @escaping (Bool) -> Void) {
        stateQueue.async {
            let msg = self.nextMessage(msg: "method", extra: [
                "method": "login",
                "params": [["resume": token]]
            ])
            
            let msgId = msg["id"] as? String
            
            self.pendingCallbacks[msgId ?? ""] = { [weak self] data in
                self?.pendingCallbacks.removeValue(forKey: msgId ?? "")
                let hasError = data["error"] != nil
                #if DEBUG
                if hasError {
                    print("[\(Self.TAG)] Login failed: \(data["error"] ?? "unknown")")
                } else {
                    print("[\(Self.TAG)] Login succeeded")
                }
                #endif
                completion(!hasError)
            }
            
            self.send(msg) { [weak self] success in
                guard let self else { return }
                if !success {
                    self.stateQueue.async {
                        guard self.pendingCallbacks.removeValue(forKey: msgId ?? "") != nil else { return }
                        completion(false)
                    }
                }
            }
        }
    }
    
    // MARK: - Subscribe
    
    func subscribe(name: String, params: [Any], completion: @escaping (Bool) -> Void) {
        stateQueue.async {
            let msg = self.nextMessage(msg: "sub", extra: [
                "name": name,
                "params": params
            ])
            
            let msgId = msg["id"] as? String
            
            self.pendingCallbacks[msgId ?? ""] = { [weak self] data in
                self?.pendingCallbacks.removeValue(forKey: msgId ?? "")
                let didSubscribe = (data["msg"] as? String) == "ready" && data["error"] == nil
                #if DEBUG
                if didSubscribe {
                    print("[\(Self.TAG)] Subscribed to \(name)")
                } else {
                    print("[\(Self.TAG)] Failed to subscribe to \(name): \(data["error"] ?? "nosub")")
                }
                #endif
                completion(didSubscribe)
            }
            
            self.send(msg) { [weak self] success in
                guard let self else { return }
                if !success {
                    self.stateQueue.async {
                        guard self.pendingCallbacks.removeValue(forKey: msgId ?? "") != nil else { return }
                        completion(false)
                    }
                }
            }
        }
    }
    
    // MARK: - Disconnect
    
    func disconnect() {
        if DispatchQueue.getSpecific(key: stateQueueKey) != nil {
            disconnectOnStateQueue(error: nil)
        } else {
            stateQueue.async { [weak self] in
                self?.disconnectOnStateQueue(error: nil)
            }
        }
    }

    private func disconnectOnStateQueue(error: Error?) {
        #if DEBUG
        print("[\(Self.TAG)] Disconnecting")
        #endif
        isConnected = false

        let pendingToFail = Array(pendingCallbacks.values)
        let connectedToFail = connectedCallback
        let queuedToFail = queuedMethodCalls

        pendingCallbacks.removeAll()
        queuedMethodCalls.removeAll()
        connectedCallback = nil
        onCollectionMessage = nil
        webSocketTask?.cancel(with: .normalClosure, reason: nil)
        webSocketTask = nil
        urlSession?.invalidateAndCancel()
        urlSession = nil

        connectedToFail?(false)
        var errorPayload: [String: Any] = ["error": ["reason": "disconnected"]]
        if let error {
            errorPayload["error"] = ["reason": "disconnected", "underlying": error.localizedDescription]
        }
        pendingToFail.forEach { $0(errorPayload) }
        queuedToFail.forEach { $0.completion(false) }
    }
    
    // MARK: - Private
    
    private var pendingCallbacks: [String: ([String: Any]) -> Void] = [:]
    private var queuedMethodCalls: [QueuedMethodCall] = []
    private var connectedCallback: ((Bool) -> Void)?
    
    private func nextMessage(msg: String, extra: [String: Any] = [:]) -> [String: Any] {
        sendCounter += 1
        var dict: [String: Any] = ["msg": msg, "id": "ddp-\(sendCounter)"]
        for (key, value) in extra {
            dict[key] = value
        }
        return dict
    }
    
    private func send(_ dict: [String: Any], completion: @escaping (Bool) -> Void) {
        guard let data = try? JSONSerialization.data(withJSONObject: dict),
              let string = String(data: data, encoding: .utf8) else {
            completion(false)
            return
        }

        guard let webSocketTask else {
            completion(false)
            return
        }
        
        webSocketTask.send(.string(string)) { error in
            if let error = error {
                #if DEBUG
                print("[\(Self.TAG)] Send error: \(error.localizedDescription)")
                #endif
                completion(false)
            } else {
                completion(true)
            }
        }
    }

    func callMethod(_ method: String, params: [Any], completion: @escaping (Bool) -> Void) {
        stateQueue.async {
            let msg = self.nextMessage(msg: "method", extra: [
                "method": method,
                "params": params
            ])

            let msgId = msg["id"] as? String

            self.pendingCallbacks[msgId ?? ""] = { [weak self] data in
                self?.pendingCallbacks.removeValue(forKey: msgId ?? "")
                let hasError = data["error"] != nil
                #if DEBUG
                if hasError {
                    print("[\(Self.TAG)] Method \(method) failed: \(data["error"] ?? "unknown")")
                }
                #endif
                completion(!hasError)
            }

            self.send(msg) { [weak self] success in
                guard let self else { return }
                if !success {
                    self.stateQueue.async {
                        guard self.pendingCallbacks.removeValue(forKey: msgId ?? "") != nil else { return }
                        completion(false)
                    }
                }
            }
        }
    }

    func queueMethodCall(_ method: String, params: [Any], completion: @escaping (Bool) -> Void = { _ in }) {
        stateQueue.async {
            self.queuedMethodCalls.append(
                QueuedMethodCall(
                    method: method,
                    params: params,
                    completion: completion
                )
            )
        }
    }

    func hasQueuedMethodCalls() -> Bool {
        if DispatchQueue.getSpecific(key: stateQueueKey) != nil {
            return !queuedMethodCalls.isEmpty
        }

        return stateQueue.sync {
            !queuedMethodCalls.isEmpty
        }
    }

    func flushQueuedMethodCalls() {
        stateQueue.async {
            let queuedCalls = self.queuedMethodCalls
            self.queuedMethodCalls.removeAll()

            queuedCalls.forEach { queuedCall in
                self.callMethod(queuedCall.method, params: queuedCall.params, completion: queuedCall.completion)
            }
        }
    }

    func clearQueuedMethodCalls() {
        if DispatchQueue.getSpecific(key: stateQueueKey) != nil {
            queuedMethodCalls.removeAll()
            return
        }

        stateQueue.async {
            self.queuedMethodCalls.removeAll()
        }
    }
    
    private func waitForConnected(timeout: TimeInterval, completion: @escaping (Bool) -> Void) {
        connectedCallback = completion
        DispatchQueue.main.asyncAfter(deadline: .now() + timeout) { [weak self] in
            self?.stateQueue.async {
                guard let self = self, let cb = self.connectedCallback else { return }
                self.connectedCallback = nil
                #if DEBUG
                print("[\(Self.TAG)] Connect timeout")
                #endif
                cb(false)
            }
        }
    }
    
    private func listenForMessages(task: URLSessionWebSocketTask) {
        task.receive { [weak self] result in
            self?.stateQueue.async { [weak self] in
                guard let self else { return }
                self.handleWebSocketReceiveResult(result, for: task)
            }
        }
    }

    private func handleWebSocketReceiveResult(_ result: Result<URLSessionWebSocketTask.Message, Error>, for task: URLSessionWebSocketTask) {
        guard task === webSocketTask else { return }

        switch result {
        case .success(let message):
            switch message {
            case .string(let text):
                handleMessage(text)
            default:
                break
            }
            guard task === webSocketTask else { return }
            listenForMessages(task: task)

        case .failure(let error):
            #if DEBUG
            print("[\(Self.TAG)] Receive error: \(error.localizedDescription)")
            #endif
            disconnectOnStateQueue(error: error)
        }
    }
    
    private func handleMessage(_ text: String) {
        guard let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return
        }
        
        let msg = json["msg"] as? String
        
        switch msg {
        case "connected":
            isConnected = true
            if let cb = connectedCallback {
                connectedCallback = nil
                cb(true)
            }
            
        case "ping":
            send(["msg": "pong"]) { _ in }
            
        case "result":
            if let id = json["id"] as? String, let cb = pendingCallbacks[id] {
                cb(json)
            }
            
        case "ready":
            if let subs = json["subs"] as? [String] {
                subs.forEach { subId in
                    pendingCallbacks[subId]?(json)
                }
            }
            
        case "changed", "added", "removed":
            if (json["collection"] as? String) != nil {
                onCollectionMessage?(json)
            }
            
        case "nosub":
            if let id = json["id"] as? String, let cb = pendingCallbacks[id] {
                cb(json)
            }
            
        default:
            if (json["collection"] as? String) != nil {
                onCollectionMessage?(json)
            }
        }
    }
    
    // MARK: - URL Helpers
    
    private static func buildWebSocketURL(host: String) -> String {
        var cleaned = host
        
        if cleaned.hasSuffix("/") {
            cleaned = String(cleaned.dropLast())
        }
        
        let useSsl: Bool
        if cleaned.hasPrefix("https://") {
            useSsl = true
            cleaned = String(cleaned.dropFirst("https://".count))
        } else if cleaned.hasPrefix("http://") {
            useSsl = false
            cleaned = String(cleaned.dropFirst("http://".count))
        } else {
            useSsl = true
        }
        
        let scheme = useSsl ? "wss" : "ws"
        return "\(scheme)://\(cleaned)/websocket"
    }
}

/// Forwards URL authentication challenges to the existing `Challenge` implementation in `SSLPinning.mm`
/// (same path as `RCTHTTPRequestHandler` swizzling) so WebSocket TLS uses the same client-certificate flow.
private final class DDPClientURLSessionChallengeDelegate: NSObject, URLSessionDelegate {
    func urlSession(
        _ session: URLSession,
        didReceive challenge: URLAuthenticationChallenge,
        completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
    ) {
        Challenge.runChallenge(session, didReceiveChallenge: challenge, completionHandler: completionHandler)
    }
}

File: android/app/src/main/java/chat/rocket/reactnative/voip/DDPClient.kt
package chat.rocket.reactnative.voip

import android.os.Handler
import android.os.Looper
import android.util.Log
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONArray
import org.json.JSONObject
import chat.rocket.reactnative.networking.SSLPinningTurboModule
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

class DDPClient {
    private data class QueuedMethodCall(
        val method: String,
        val params: JSONArray,
        val callback: (Boolean) -> Unit
    )

    companion object {
        private const val TAG = "RocketChat.DDPClient"
        private val sharedClient: OkHttpClient by lazy {
            SSLPinningTurboModule.getSharedOkHttpClient() ?: OkHttpClient.Builder()
                .pingInterval(30, TimeUnit.SECONDS)
                .build()
        }
    }

    private var webSocket: WebSocket? = null
    private val client: OkHttpClient = sharedClient
    private val sendCounter = AtomicInteger(0)

    @Volatile
    private var isConnected = false

    private val mainHandler = Handler(Looper.getMainLooper())

    private val pendingCallbacks = mutableMapOf<String, (JSONObject) -> Unit>()
    private val queuedMethodCalls = mutableListOf<QueuedMethodCall>()

    @Volatile
    private var connectedCallback: ((Boolean) -> Unit)? = null

    @Volatile
    private var connectTimeoutRunnable: Runnable? = null

    private val connectResultDelivered = AtomicBoolean(false)

    var onCollectionMessage: ((JSONObject) -> Unit)? = null

    fun connect(host: String, callback: (Boolean) -> Unit) {
        resetConnectHandshakeState()

        val wsUrl = buildWebSocketURL(host)

        Log.d(TAG, "Connecting to $wsUrl")

        val request = Request.Builder().url(wsUrl).build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                Log.d(TAG, "WebSocket opened")
                val connectMsg = JSONObject().apply {
                    put("msg", "connect")
                    put("version", "1")
                    put("support", JSONArray().apply {
                        put("1"); put("pre2"); put("pre1")
                    })
                }
                webSocket.send(connectMsg.toString())
                waitForConnected(10_000L, callback)
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                if (webSocket !== this@DDPClient.webSocket) return
                handleMessage(text)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                if (webSocket !== this@DDPClient.webSocket) return
                Log.e(TAG, "WebSocket failure: ${t.message}")
                isConnected = false
                mainHandler.post {
                    // Re-check identity on main thread: disconnect()+connect() can interleave
                    // between the outer guard (OkHttp thread) and this runnable's execution.
                    // Without this re-check, a stale failure can hijack a newly installed
                    // connectedCallback via the CAS in tryDeliverConnectOutcome.
                    if (webSocket !== this@DDPClient.webSocket) return@post
                    tryDeliverConnectOutcome(false)
                }
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                if (webSocket !== this@DDPClient.webSocket) return
                Log.d(TAG, "WebSocket closed: $code $reason")
                isConnected = false
            }
        })
    }

    fun login(token: String, callback: (Boolean) -> Unit) {
        val msg = nextMessage("method").apply {
            put("method", "login")
            put("params", JSONArray().apply {
                put(JSONObject().apply { put("resume", token) })
            })
        }

        val msgId = msg.getString("id")

        synchronized(pendingCallbacks) {
            pendingCallbacks[msgId] = { data ->
                synchronized(pendingCallbacks) { pendingCallbacks.remove(msgId) }
                val hasError = data.has("error")
                if (hasError) {
                    Log.e(TAG, "Login failed: ${data.opt("error")}")
                } else {
                    Log.d(TAG, "Login succeeded")
                }
                mainHandler.post { callback(!hasError) }
            }
        }

        if (!send(msg)) {
            mainHandler.post { callback(false) }
        }
    }

    fun subscribe(name: String, params: JSONArray, callback: (Boolean) -> Unit) {
        val msg = nextMessage("sub").apply {
            put("name", name)
            put("params", params)
        }

        val msgId = msg.getString("id")

        synchronized(pendingCallbacks) {
            pendingCallbacks[msgId] = { data ->
                synchronized(pendingCallbacks) { pendingCallbacks.remove(msgId) }
                val didSubscribe = data.optString("msg") == "ready" && !data.has("error")
                if (didSubscribe) {
                    Log.d(TAG, "Subscribed to $name")
                } else {
                    Log.e(TAG, "Failed to subscribe to $name: ${data.opt("error") ?: "nosub"}")
                }
                mainHandler.post { callback(didSubscribe) }
            }
        }

        if (!send(msg)) {
            synchronized(pendingCallbacks) { pendingCallbacks.remove(msgId) }
            mainHandler.post { callback(false) }
        }
    }

    fun disconnect() {
        Log.d(TAG, "Disconnecting")
        isConnected = false
        cancelConnectTimeout()
        synchronized(pendingCallbacks) { pendingCallbacks.clear() }
        clearQueuedMethodCalls()
        connectedCallback = null
        onCollectionMessage = null
        webSocket?.close(1000, null)
        webSocket = null
    }

    private fun nextMessage(msg: String): JSONObject {
        val nextId = sendCounter.incrementAndGet()
        return JSONObject().apply {
            put("msg", msg)
            put("id", "ddp-$nextId")
        }
    }

    private fun send(json: JSONObject): Boolean {
        val ws = webSocket ?: return false
        return ws.send(json.toString())
    }

    fun callMethod(method: String, params: JSONArray, callback: (Boolean) -> Unit) {
        val msg = nextMessage("method").apply {
            put("method", method)
            put("params", params)
        }

        val msgId = msg.getString("id")

        synchronized(pendingCallbacks) {
            pendingCallbacks[msgId] = { data ->
                synchronized(pendingCallbacks) { pendingCallbacks.remove(msgId) }
                val hasError = data.has("error")
                if (hasError) {
                    Log.e(TAG, "Method $method failed: ${data.opt("error")}")
                }
                mainHandler.post { callback(!hasError) }
            }
        }

        if (!send(msg)) {
            synchronized(pendingCallbacks) { pendingCallbacks.remove(msgId) }
            mainHandler.post { callback(false) }
        }
    }

    fun queueMethodCall(method: String, params: JSONArray, callback: (Boolean) -> Unit = {}) {
        synchronized(queuedMethodCalls) {
            queuedMethodCalls.add(
                QueuedMethodCall(
                    method = method,
                    params = params,
                    callback = callback
                )
            )
        }
    }

    fun hasQueuedMethodCalls(): Boolean =
        synchronized(queuedMethodCalls) { queuedMethodCalls.isNotEmpty() }

    fun flushQueuedMethodCalls() {
        val queuedCalls = synchronized(queuedMethodCalls) {
            queuedMethodCalls.toList().also { queuedMethodCalls.clear() }
        }

        queuedCalls.forEach { queuedCall ->
            callMethod(queuedCall.method, queuedCall.params, queuedCall.callback)
        }
    }

    fun clearQueuedMethodCalls() {
        synchronized(queuedMethodCalls) {
            queuedMethodCalls.clear()
        }
    }

    private fun resetConnectHandshakeState() {
        connectResultDelivered.set(false)
    }

    /**
     * Delivers at most one outcome for the WebSocket connect handshake ([connect] callback).
     * Uses [AtomicBoolean] so [onFailure], the connect-timeout runnable, and `"connected"` cannot
     * all report conflicting results.
     */
    private fun tryDeliverConnectOutcome(success: Boolean, connectTimeout: Boolean = false) {
        if (!connectResultDelivered.compareAndSet(false, true)) {
            return
        }
        cancelConnectTimeout()
        val cb = connectedCallback
        connectedCallback = null
        if (connectTimeout) {
            Log.e(TAG, "Connect timeout")
        }
        mainHandler.post {
            cb?.invoke(success)
        }
    }

    private fun waitForConnected(timeoutMs: Long, callback: (Boolean) -> Unit) {
        connectedCallback = callback
        cancelConnectTimeout()
        val runnable = Runnable {
            connectTimeoutRunnable = null
            tryDeliverConnectOutcome(false, connectTimeout = true)
        }
        connectTimeoutRunnable = runnable
        mainHandler.postDelayed(runnable, timeoutMs)
    }

    private fun cancelConnectTimeout() {
        connectTimeoutRunnable?.let { mainHandler.removeCallbacks(it) }
        connectTimeoutRunnable = null
    }

    private fun handleMessage(text: String) {
        val json = try {
            JSONObject(text)
        } catch (e: Exception) {
            return
        }

        when (json.optString("msg")) {
            "connected" -> {
                isConnected = true
                tryDeliverConnectOutcome(true)
            }

            "ping" -> {
                send(JSONObject().apply { put("msg", "pong") })
            }

            "result" -> {
                val id = json.optString("id")
                val cb = synchronized(pendingCallbacks) { pendingCallbacks[id] }
                cb?.invoke(json)
            }

            "ready" -> {
                val subs = json.optJSONArray("subs")
                if (subs != null) {
                    for (index in 0 until subs.length()) {
                        val subId = subs.optString(index)
                        if (subId.isEmpty()) {
                            continue
                        }

                        val cb = synchronized(pendingCallbacks) { pendingCallbacks[subId] }
                        cb?.invoke(json)
                    }
                }
            }

            "changed", "added", "removed" -> {
                onCollectionMessage?.invoke(json)
            }

            "nosub" -> {
                val id = json.optString("id")
                val cb = synchronized(pendingCallbacks) { pendingCallbacks[id] }
                cb?.invoke(json)
            }

            else -> {
                if (json.has("collection")) {
                    onCollectionMessage?.invoke(json)
                }
            }
        }
    }

    private fun buildWebSocketURL(host: String): String {
        var normalizedHost = host.trimEnd('/')

        val useSsl: Boolean
        when {
            normalizedHost.startsWith("https://") -> {
                useSsl = true
                normalizedHost = normalizedHost.removePrefix("https://")
            }
            normalizedHost.startsWith("http://") -> {
                throw IllegalStateException("DDPClient does not support plaintext http:// servers — use https://")
            }
            else -> {
                useSsl = true
            }
        }

        val scheme = if (useSsl) "wss" else "ws"
        return "$scheme://$normalizedHost/websocket"
    }

    internal fun testStartConnectTimeout(timeoutMs: Long, callback: (Boolean) -> Unit) {
        resetConnectHandshakeState()
        waitForConnected(timeoutMs, callback)
    }

    internal fun testDeliverRawMessage(text: String) {
        handleMessage(text)
    }

    internal fun testDeliverConnectFailure(fromWebSocket: WebSocket? = null) {
        mainHandler.post {
            if (fromWebSocket != null && fromWebSocket !== this@DDPClient.webSocket) return@post
            tryDeliverConnectOutcome(false)
        }
    }

    internal fun testSetActiveWebSocket(ws: WebSocket?) {
        this.webSocket = ws
    }
}
